### DISCUSSION TITLE
How to supply a VAULT_NAMESPACE env to Kustomize Controller when decrypting using Hashicorp Vault

### DISCUSSION THREAD

**Discussion by sst1xx** [fluxcd/flux2#3365] [Q&A] [answered]
Hi,

that's could be a simple question but I could not find how to pass the essential env  "VAULT_NAMESPACE" for decryption in cluster.
We are using Vault in multi-tenants setup and have to export this variable before any actions with vault. And this variable should be  different for every ks8 namespace.

It's possible to encrypt data locally but how to set this variable inside cluster for application? sops.yaml doesn't have such fields.
Does somebody have the same problem?

Thanks.

**Comment by jaythamke** [comment]
@sst1xx did you find any work around?
@stefanprodan  Could you please reply on this issue? I am also facing this problem.
Additionally, I do not want to provide global env variable 'VAULT_NAMESPACE' for kustomize controller. 
I posted the query in the Slack channel as well https://kubernetes.slack.com/archives/CBT6N1ASG/p1698612591329179

Thanks

**Comment by souleb** [comment, marked as answer]
We use `Sops` to handle `HashiCorp Vault` secrets. Unfortunately Sops doesn't support `VAULT_NAMESPACE`. I would suggest you fill in a feature request in `https://github.com/getsops`.

I believe the following function has to be adapted to support setting a header with `"X-Vault-Namespace"`:
https://github.com/getsops/sops/blob/d7c2d7d30f1e3991c8646c1ad829a1c34263e05c/hcvault/keysource.go#L285

The value has to be part of the secret in order to be used for decryption too.

**Comment by jaythamke** [comment]
Thanks for reply,
I have different opinion than sops supporting this feature. Why vault-token secret does not accept vault namespace along with vault token? Something like following as extra key - `vault-ns`

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: sops-hcvault
  namespace: ns01
data:
  sops.vault-token: <base64-enc-token>
  vault-ns: <base64-enc-vault-namespace> 
```

**Reply by stefanprodan** [reply]
The Vault metadata is encoded by SOPS in the secret. Flux should extract the namespace from the metadata set by SOPS, in the same way it extracts the Vault URL.

**Reply by jaythamke** [reply]
Thanks for taking out some time to respond,
In which K8s secret's metadata (label/annotation and keys) I have to provide vault namespace? Do you have the documentation or point to source code? Or could you please extend above K8s secret manifest? Thanks

**Reply by stefanprodan** [reply]
This needs to be implemented in [SOPS](https://github.com/getsops/sops). Flux uses SOPS as a library, once SOPS encodes the namespace in the metadata, Flux will know about it.
